train_and_eval/train/train_imagenet.py

```diff
@@ -0,0 +1,607 @@
+    log_level=Param(int, '0 if only at end 1 otherwise', default=1)
+)
+
+Section('validation', 'Validation parameters stuff').params(
+    batch_size=Param(int, 'The batch size for validation', default=512),
+    resolution=Param(int, 'final resized validation image size', default=224),
+    lr_tta=Param(int, 'should do lr flipping/avging at test time', default=1)
+)
+
+Section('training', 'training hyper param stuff').params(
+    eval_only=Param(int, 'eval only?', default=0),
+    path=Param(str, 'weight path for trained model', default=None),
+    batch_size=Param(int, 'The batch size', default=512),
+    optimizer=Param(And(str, OneOf(['sgd'])), 'The optimizer', default='sgd'),
+    momentum=Param(float, 'SGD momentum', default=0.9),
+    weight_decay=Param(float, 'weight decay', default=4e-5),
+    epochs=Param(int, 'number of epochs', default=30),
+    label_smoothing=Param(float, 'label smoothing parameter', default=0.1),
+    distributed=Param(int, 'is distributed?', default=0),
+    use_blurpool=Param(int, 'use blurpool?', default=0)
+)
+
+Section('dist', 'distributed training options').params(
+    world_size=Param(int, 'number gpus', default=1),
+    address=Param(str, 'address', default='localhost'),
+    port=Param(str, 'port', default='12355')
+)
+
+IMAGENET_MEAN = np.array([0.485, 0.456, 0.406]) * 255
+IMAGENET_STD = np.array([0.229, 0.224, 0.225]) * 255
+DEFAULT_CROP_RATIO = 224/256
+
+@param('lr.lr')
+@param('lr.step_ratio')
+@param('lr.step_length')
+@param('training.epochs')
+def get_step_lr(epoch, lr, step_ratio, step_length, epochs):
+    if epoch >= epochs:
+        return 0
+
+    num_steps = epoch // step_length
+    return step_ratio**num_steps * lr
+
+@param('lr.lr')
+def get_constant_lr(epoch, lr):
+    return lr
+
+@param('lr.lr')
+@param('training.epochs')
+@param('lr.lr_peak_epoch')
+def get_cyclic_lr(epoch, lr, epochs, lr_peak_epoch):
+    xs = [0, lr_peak_epoch, epochs]
+    ys = [1e-4 * lr, lr, 0]
+    return np.interp([epoch], xs, ys)[0]
+
+class BlurPoolConv2d(ch.nn.Module):
+    def __init__(self, conv):
+        super().__init__()
+        default_filter = ch.tensor([[[[1, 2, 1], [2, 4, 2], [1, 2, 1]]]]) / 16.0
+        filt = default_filter.repeat(conv.in_channels, 1, 1, 1)
+        self.conv = conv
+        self.register_buffer('blur_filter', filt)
+
+    def forward(self, x):
+        blurred = F.conv2d(x, self.blur_filter, stride=1, padding=(1, 1),
+                           groups=self.conv.in_channels, bias=None)
+        return self.conv.forward(blurred)
+
+class ImageNetTrainer:
+    @param('training.distributed')
+    @param('model.efficient')
+    @param('model.mrl')
+    @param('model.nesting_start')
+    @param('model.fixed_feature')
+    def __init__(self, gpu, distributed, efficient, mrl, nesting_start, fixed_feature):
+        self.all_params = get_current_config(); 
+        self.gpu = gpu
+        self.efficient = efficient
+        self.mrl = mrl
```

Comment:
delete: `self.mrl` is not being used anywhere besides to set `self.nesting`

Reply:
Noted.